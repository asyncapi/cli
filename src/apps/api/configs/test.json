{
  "env": "test",
  "log": {
    "format": "dev",
    "dir": "../logs"
  },
  "cors": {
    "origin": true,
    "credentials": false
  },
  "request": {
    "body": {
      "limit": "5mb"
    }
  }
}
