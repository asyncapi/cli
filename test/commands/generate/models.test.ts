/* eslint-disable sonarjs/no-duplicate-string */
/* eslint-disable sonarjs/no-identical-functions */
import path from 'path';
import { test } from '@oclif/test';
const generalOptions = ['generate:models'];
const outputDir = './test/commands/generate/models';

describe('models', () => {
  test
    .stderr()
    .stdout()
    .command([...generalOptions, 'typescript', 'http://bit.ly/asyncapi'])
    .it('works with remote AsyncAPI files', (ctx, done) => {
      expect(ctx.stderr).toEqual('');
      expect(ctx.stdout).toMatchSnapshot();
      done();
    });

  test
    .stderr()
    .stdout()
    .command([...generalOptions, 'random', './test/specification.yml', `-o=${ path.resolve(outputDir, './random')}`])
    .it('fails when it dont know the language', (ctx, done) => {
      expect(ctx.stderr).toEqual('Error: Expected random to be one of: typescript, csharp, golang, java, javascript, dart, python, rust\nSee more help with --help\n');
      expect(ctx.stdout).toEqual('');
      done();
    });

  test
    .stderr()
    .stdout()
    .command([...generalOptions, 'typescript', './test/specification.yml'])
    .it('works when generating in memory', (ctx, done) => {
      expect(ctx.stderr).toEqual('');
      expect(ctx.stdout).toMatchSnapshot();
      done();
    });

  describe('for TypeScript', () => {
    test
      .stderr()
      .stdout()
      .command([...generalOptions, 'typescript', './test/specification.yml', `-o=${ path.resolve(outputDir, './ts')}`])
      .it('works when file path is passed', (ctx, done) => {
        expect(ctx.stderr).toEqual('');
        expect(ctx.stdout).toContain(
          'Successfully generated the following models: '
        );
        done();
      });
    test
      .stderr()
      .stdout()
      .command([...generalOptions, 'typescript', './test/specification.yml', `-o=${ path.resolve(outputDir, './ts')}`, '--tsJsonBinPack'])
      .it('works when tsJsonBinPack is set', (ctx, done) => {
        expect(ctx.stderr).toEqual('');
        expect(ctx.stdout).toContain(
          'Successfully generated the following models: '
        );
        done();
      });
  });

  describe('for JavaScript', () => {
    test
      .stderr()
      .stdout()
      .command([...generalOptions, 'javascript', './test/specification.yml', `-o=${ path.resolve(outputDir, './js')}`])
      .it('works when file path is passed', (ctx, done) => {
        expect(ctx.stdout).toContain(
          'Successfully generated the following models: '
        );
        expect(ctx.stderr).toEqual('');
        done();
      });
  });

  describe('for Python', () => {
    test
      .stderr()
      .stdout()
      .command([...generalOptions, 'python', './test/specification.yml', `-o=${ path.resolve(outputDir, './python')}`])
      .it('works when file path is passed', (ctx, done) => {
        expect(ctx.stdout).toContain(
          'Successfully generated the following models: '
        );
        expect(ctx.stderr).toEqual('');
        done();
      });
  });

  describe('for Rust', () => {
    test
      .stderr()
      .stdout()
      .command([...generalOptions, 'rust', './test/specification.yml', `-o=${ path.resolve(outputDir, './rust')}`])
      .it('works when file path is passed', (ctx, done) => {
        expect(ctx.stdout).toContain(
          'Successfully generated the following models: '
        );
        expect(ctx.stderr).toEqual('');
        done();
      });
  });

  describe('for C#', () => {
    test
      .stderr()
      .stdout()
      .command([...generalOptions, 'csharp', './test/specification.yml', `-o=${path.resolve(outputDir, './csharp')}`, '--namespace=\'asyncapi.models\''])
      .it('works when file path is passed', (ctx, done) => {
        expect(ctx.stderr).toEqual('');
        expect(ctx.stdout).toContain(
          'Successfully generated the following models: '
        );
        done();
      });
    test
      .stderr()
      .stdout()
      .command([...generalOptions, 'csharp', './test/specification.yml', `-o=${ path.resolve(outputDir, './csharp')}`])
      .it('fails when no namespace provided', (ctx, done) => {
        expect(ctx.stderr).toEqual('Error: In order to generate models to C#, we need to know which namespace they are under. Add `--namespace=NAMESPACE` to set the desired namespace.\n');
        expect(ctx.stdout).toEqual('');
        done();
      });
    test
      .stderr()
      .stdout()
      .command([...generalOptions, 'csharp', './test/specification.yml', `-o=${ path.resolve(outputDir, './csharp')}`, '--namespace=\'asyncapi.models\'', '--csharpAutoImplement'])
      .it('works when auto implement properties flag is passed', (ctx, done) => {
        expect(ctx.stderr).toEqual('');
        expect(ctx.stdout).toContain(
          'Successfully generated the following models: '
        );
        done();
      });
    test
      .stderr()
      .stdout()
      .command([...generalOptions, 'csharp', './test/specification.yml', `-o=${ path.resolve(outputDir, './csharp')}`, '--namespace=\'asyncapi.models\'', '--csharpArrayType=List']) 
      .it('works when array type is provided', (ctx, done) => {
        expect(ctx.stderr).toEqual('');
        expect(ctx.stdout).toContain(
          'Successfully generated the following models: '
        );
        done();
      });
  });

  describe('for Java', () => {
    test
      .stderr()
      .stdout()
      .command([...generalOptions, 'java', './test/specification.yml', `-o=${ path.resolve(outputDir, './java')}`, '--packageName', 'test.pkg'])
      .it('works when file path is passed', (ctx, done) => {
        expect(ctx.stderr).toEqual('');
        expect(ctx.stdout).toContain(
          'Successfully generated the following models: '
        );
        done();
      });
    test
      .stderr()
      .stdout()
      .command([...generalOptions, 'java', './test/specification.yml', `-o=${ path.resolve(outputDir, './java')}`])
      .it('fails when no package defined', (ctx, done) => {
        expect(ctx.stderr).toEqual('Error: In order to generate models to Java, we need to know which package they are under. Add `--packageName=PACKAGENAME` to set the desired package name.\n');
        expect(ctx.stdout).toEqual('');
        done();
      });
  });

  describe('for Go', () => {
    test
      .stderr()
      .stdout()
      .command([...generalOptions, 'golang', './test/specification.yml', `-o=${ path.resolve(outputDir, './go')}`, '--packageName', 'asyncapi.models'])
      .it('works when file path is passed', (ctx, done) => {
        expect(ctx.stderr).toEqual('');
        expect(ctx.stdout).toContain(
          'Successfully generated the following models: '
        );
        done();
      });
    test
      .stderr()
      .stdout()
      .command([...generalOptions, 'golang', './test/specification.yml', `-o=${ path.resolve(outputDir, './go')}`])
      .it('fails when no package defined', (ctx, done) => {
        expect(ctx.stderr).toEqual('Error: In order to generate models to Go, we need to know which package they are under. Add `--packageName=PACKAGENAME` to set the desired package name.\n');
        expect(ctx.stdout).toEqual('');
        done();
      });
  });

  describe('for Dart', () => {
    test
      .stderr()
      .stdout()
      .command([...generalOptions, 'dart', './test/specification.yml', `-o=${ path.resolve(outputDir, './dart')}`, '--packageName', 'asyncapi.models'])
      .it('works when file path is passed', (ctx, done) => {
        expect(ctx.stderr).toEqual('');
        expect(ctx.stdout).toContain(
          'Successfully generated the following models: '
        );
        done();
      });
    test
      .stderr()
      .stdout()
      .command([...generalOptions, 'dart', './test/specification.yml', `-o=${ path.resolve(outputDir, './dart')}`])
      .it('fails when no package defined', (ctx, done) => {
        expect(ctx.stderr).toEqual('Error: In order to generate models to Dart, we need to know which package they are under. Add `--packageName=PACKAGENAME` to set the desired package name.\n');
        expect(ctx.stdout).toEqual('');
        done();
      });
  });

  describe('with logging diagnostics', () => {
    test
      .stderr()
      .stdout()
      .command([...generalOptions, 'typescript', 'http://bit.ly/asyncapi', '--log-diagnostics'])
      .it('works with remote AsyncAPI files', (ctx, done) => {
        expect(ctx.stderr).toEqual('');
        expect(ctx.stdout).toMatch('URL http://bit.ly/asyncapi is valid but has (itself and/or referenced documents) governance issues.');
        done();
      });
  });
});
