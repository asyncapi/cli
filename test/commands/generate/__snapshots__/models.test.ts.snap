// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`models works when generating in memory 1`] = `
"Successfully generated the following models: 
## Model name: AnonymousSchema_1

class AnonymousSchema_1 {
  private _displayName?: string;
  private _email?: string;
  private _additionalProperties?: Map<string, any>;

  constructor(input: {
    displayName?: string,
    email?: string,
    additionalProperties?: Map<string, any>,
  }) {
    this._displayName = input.displayName;
    this._email = input.email;
    this._additionalProperties = input.additionalProperties;
  }

  get displayName(): string | undefined { return this._displayName; }
  set displayName(displayName: string | undefined) { this._displayName = displayName; }

  get email(): string | undefined { return this._email; }
  set email(email: string | undefined) { this._email = email; }

  get additionalProperties(): Map<string, any> | undefined { return this._additionalProperties; }
  set additionalProperties(additionalProperties: Map<string, any> | undefined) { this._additionalProperties = additionalProperties; }
}
export default AnonymousSchema_1;


"
`;

exports[`models works with remote AsyncAPI files 1`] = `
"Successfully generated the following models: 
## Model name: LightMeasuredPayload

class LightMeasuredPayload {
  private _lumens?: number;
  private _sentAt?: string;
  private _additionalProperties?: Map<string, any>;

  constructor(input: {
    lumens?: number,
    sentAt?: string,
    additionalProperties?: Map<string, any>,
  }) {
    this._lumens = input.lumens;
    this._sentAt = input.sentAt;
    this._additionalProperties = input.additionalProperties;
  }

  get lumens(): number | undefined { return this._lumens; }
  set lumens(lumens: number | undefined) { this._lumens = lumens; }

  get sentAt(): string | undefined { return this._sentAt; }
  set sentAt(sentAt: string | undefined) { this._sentAt = sentAt; }

  get additionalProperties(): Map<string, any> | undefined { return this._additionalProperties; }
  set additionalProperties(additionalProperties: Map<string, any> | undefined) { this._additionalProperties = additionalProperties; }
}
export default LightMeasuredPayload;



## Model name: TurnOnOffPayload
import AnonymousSchema_6 from './AnonymousSchema_6';
class TurnOnOffPayload {
  private _command?: AnonymousSchema_6;
  private _sentAt?: string;
  private _additionalProperties?: Map<string, any>;

  constructor(input: {
    command?: AnonymousSchema_6,
    sentAt?: string,
    additionalProperties?: Map<string, any>,
  }) {
    this._command = input.command;
    this._sentAt = input.sentAt;
    this._additionalProperties = input.additionalProperties;
  }

  get command(): AnonymousSchema_6 | undefined { return this._command; }
  set command(command: AnonymousSchema_6 | undefined) { this._command = command; }

  get sentAt(): string | undefined { return this._sentAt; }
  set sentAt(sentAt: string | undefined) { this._sentAt = sentAt; }

  get additionalProperties(): Map<string, any> | undefined { return this._additionalProperties; }
  set additionalProperties(additionalProperties: Map<string, any> | undefined) { this._additionalProperties = additionalProperties; }
}
export default TurnOnOffPayload;



## Model name: AnonymousSchema_6

enum AnonymousSchema_6 {
  ON = "on",
  OFF = "off",
}
export default AnonymousSchema_6;



## Model name: DimLightPayload

class DimLightPayload {
  private _percentage?: number;
  private _sentAt?: string;
  private _additionalProperties?: Map<string, any>;

  constructor(input: {
    percentage?: number,
    sentAt?: string,
    additionalProperties?: Map<string, any>,
  }) {
    this._percentage = input.percentage;
    this._sentAt = input.sentAt;
    this._additionalProperties = input.additionalProperties;
  }

  get percentage(): number | undefined { return this._percentage; }
  set percentage(percentage: number | undefined) { this._percentage = percentage; }

  get sentAt(): string | undefined { return this._sentAt; }
  set sentAt(sentAt: string | undefined) { this._sentAt = sentAt; }

  get additionalProperties(): Map<string, any> | undefined { return this._additionalProperties; }
  set additionalProperties(additionalProperties: Map<string, any> | undefined) { this._additionalProperties = additionalProperties; }
}
export default DimLightPayload;


"
`;
